### DISCUSSION TITLE
About dimension of NDList when input to model

### DISCUSSION THREAD

**Discussion by freesia-sherry** [deepjavalibrary/djl#1377] [Q&A] [answered]
I met a problem that the NDList will add one more dimension to the NDArray when input into models. The expected size of inputs is 1 x 2, but when change NDArray to NDList, the size will be 1 x 1 x 2. But in my python code, i will manipulate the dimension 1 of input tensors, so it result errors. And if I use NDList save one more NDArray, it will happen errors like "received too more args", it seems that the size of NDList equals the number of inputs to model. So in that case, how can I solve my problem? Thank you!
![image](https://user-images.githubusercontent.com/25262683/142362964-ef8f78f0-beb6-43eb-b890-bbadc84cc916.png)

```java
Translator<NDArray, NDArray> translator = new Translator<NDArray, NDArray>() {
            @Override
            public Batchifier getBatchifier() {
                return Batchifier.STACK;
            }

            @Override
            public NDArray processOutput(TranslatorContext translatorContext, NDList ndList) throws Exception {
                return ndList.get(0);
            }

            @Override
            public NDList processInput(TranslatorContext translatorContext, NDArray ndArray) throws Exception {

                return new NDList(ndArray);
            }
        };

Predictor<NDArray, NDArray> predictor = model.newPredictor(translator);
        NDManager manager = NDManager.newBaseManager();
        NDArray samples = manager.create(new float[][] {{0, 1}});
        NDArray result = predictor.predict(samples);
        System.out.println(result);
```

**Comment by frankfliu** [comment]
@freesia-sherry Your model doesn't support batch, but your Translator returns `Batchifier.STACK`.

When you return `Batchifier.STACK`, DJL will create a batch dimension for you automatically. What you need to do is `return null` or simply extends `NoBatchifyTranslator`

**Reply by freesia-sherry** [reply]
Thanks for the reply. By setting return null solved my problems. But I think my model support batch.  This is my model's output, the size is 2 x 1, the batch size is 2.
![image](https://user-images.githubusercontent.com/25262683/142375175-ff464448-674b-401d-9bf6-da9b3bbeff0b.png)
In my case, it seems that one element of NDList is batch data. It might differ from what Batchifier means, so in the DJL, how to use the Batchifier correctly? Such as how to construct the input data.

**Comment by frankfliu** [comment, marked as answer]
@freesia-sherry Usually the first dimension is batch dimension, if the batchsize is 2, the input shape should be 2 x 1, not 1 x 2.

But in any case, you can always manually batch your inputs and use  `NoBatchifyTranslator` so DJL won't try to auto batch for you.

**Reply by freesia-sherry** [reply]
Thanks, I got the batch size wrong, the batch size of the case is 1, not 2. By the way, I found another function **batchPredict**, which can receive a list as inputs, each element in the list is one sample(vector), and the whole list might be batch data. Maybe I should use **batchPredict** to achieve batch prediction instead of predict.
